feat: non-blocking libdrop event handling

gbaranski opened the pull request:
~~I saw some ideas to implement a channel, but is it actually necessary?~~

~~The implementation of OnEvent() is using a mutex anyways, so no data races should be happening, and Go-routines seem to be light enough to handle this without overusing resources.~~

gbaranski commented:
I've implemented both `SyncEvent()` and `AsyncEvent()`.
Right now `SyncEvent()` is something that will be used only in tests.
`AsyncEvent()` is the method used in production code.

`OnEvent()` method has been changed to `onEvent()`, in order to prevent data races, which can happen if one uses OnEvent, and the other uses the channel.

devzbysiu commented on fileshare/event_manager.go:
```diff
@@ -127,10 +174,7 @@ func (em *EventManager) DisableNotifications() error {
-	em.mutex.Lock()
-	defer em.mutex.Unlock()
-
+func (em *EventManager) handleEvent(event Event) {
```
Doc comment still contains old name

devzbysiu commented on fileshare/event_manager.go:
```diff
@@ -66,17 +70,60 @@ func NewEventManager(
+	em.asyncEvents <- event
+}
+
+func (em *EventManager) SyncEvent(event ...Event) {
```
So this is used only in tests?

gbaranski replied:
Basically yes, but I don't want to put it to test-only file, because it can be used in production code as well.

devzbysiu replied:
I think I would go with YAGNI rule here :smile:, because we don't have any signals that synchronous processing will be needed. IMHO it's better to not try to foresee the future, it works in very rare cases

What do you think?

gbaranski replied:
@devzbysiu but then what do you do with the need for having synchronous processing in tests? Maybe I can just make the `SyncEvent()` function private?

gbaranski commented on fileshare/event_manager.go:
```diff
@@ -66,17 +70,60 @@ func NewEventManager(
+		select {
+		case e, ok := <-em.asyncEvents:
+			if !ok {
+				log.Println(internal.WarningPrefix + " asyncEvents channel closed")
```
Copilot actually suggested me this, and I guess it's because of this: 
https://github.com/NordSecurity/nordvpn-linux/blob/0e0e77eb68695f98469ee6dfe8dc21478d923ebc/fileshare/event_manager.go#L179

Will fix

gbaranski commented on fileshare/event_manager.go:
```diff
@@ -601,7 +659,7 @@ func (em *EventManager) Subscribe(id string) <-chan TransferProgressInfo {
 	defer em.mutex.Unlock()
 
-	em.transferSubscriptions[id] = make(chan TransferProgressInfo)
+	em.transferSubscriptions[id] = make(chan TransferProgressInfo, 32) // use buffered channels, because we don't want to block the event processing
```
well I decided that 32 is big enough to probably cover all the needs.
 
Maybe for further optimizations it'd be cool to actually measure what's the need during a normal file upload.

But considering that before it was an unbuffered channel, anything higher than 1 would be enough.

gbaranski commented on fileshare/event_manager.go:
```diff
@@ -66,17 +70,60 @@ func NewEventManager(
 		defaultDownloadDir:    defaultDownloadDir,
+		syncEvents:            make(chan []Event),
+		syncDoneCh:            make(chan struct{}),
+		asyncEvents:           make(chan []Event, 2048),
```
that's valid point. I set it to 2048 for tests.
I'll change it to 32, 2048 seems like way too much.

devzbysiu commented on fileshare/event_manager.go:
```diff
@@ -66,15 +70,69 @@ func NewEventManager(
+	}
+}
+
+// sends an event to the event manager
```
Make sure the code is aligned with go doc comments convention - they should start with the name of the symbol being documented

gbaranski commented:
I've removed `SyncEvent` and renamed `AsyncEvent` to `Event`. The tests are now using mockNotifier for synchronizing, and getting know if the event was processed.

devzbysiu commented on fileshare/event_manager_test.go:
```diff
@@ -41,12 +41,35 @@ type mockNotification struct {
+	}()
+
+	select {
+	case <-timeout:
```
Can we use `<-time.After(1 * time.Second)` here? I think you won't need to spawn separate goroutine then

devzbysiu commented on fileshare/event_manager.go:
```diff
@@ -206,6 +240,29 @@ func (em *EventManager) handleRequestReceivedEvent(event EventKindRequestReceive
+		case ch <- progress:
+		default:
+			log.Println(internal.WarningPrefix, " progress channel is full. removing oldest item and sending")
+			<-ch
```
So this will drop progress events in case the channel is full. I assume this will happen only when we are too slow with processing events from libdrop right?

gbaranski replied:
yes, that's right. I made an assumption that the progress events are not that important, and dropping an event is not a big deal, because you'll have other 32 events waiting for you anyways.

gbaranski commented on fileshare/events.go:
```diff
@@ -9,7 +9,7 @@ import (
 
 type EventCallback interface {
-	OnEvent(event Event)
+	Event(event ...Event)
```
yes, it is. Look at the tests.

gbaranski commented on fileshare/event_manager_test.go:
```diff
@@ -1363,3 +1374,127 @@ func TestAutoaccept(t *testing.T) {
+					em.Event(chunk...)
+				}
+				lastProgress := uint32(0)
+				for i := uint64(0); i < numEvents-2; i++ {
```
-1 because of table counting starts at 0, -1 because I want to go one number higher than the limit.

gbaranski replied:
nevermind, that was not the reason.

it was because the data is chunked improperly.

now it's fixed, and comments were added.

mariusSincovici commented on fileshare/event_manager_test.go:
```diff
@@ -1363,3 +1374,132 @@ func TestAutoaccept(t *testing.T) {
+
+	// subtract by the existing RequestQueued and FileStarted events
+	// to make sure we have the correct number of events
+	// otherwise it's not chunked properly, as the size is not disivable by chunkSize(1026 % 8 = 2)
```
[spelling] `disivable`
Add an assert into the test if the `events` length is not exactly divisible with `8`.
So if the tests is changed in the future to know sooner why it would fail.
